# di-SNE
`di-SNE` is a tool to perform t-SNE, t-distributed stochastic neighbor embedding, on a pre-filtered and clustered dataset. t-SNE aids with non-linear dimensionality reduction for high-dimensional data, making it applicable when analyzing single-cell RNA sequencing (scRNA-seq) data, as well as many other types of datasets.   
`di-SNE` utilizes the existing Scanpy library to perform clustering on the given dataset, and uses the Matplotlib library to visualize the data. A full list of required packages can be found in `tests/requirements.txt`.   

For more information on t-SNE, visit [Laurens van der Maaten's website](https://lvdmaaten.github.io/tsne/).

# Install Instructions
Installation requires the `pandas`, `numpy`, `scanpy`, and `matplotlib` (seaborn?) libraries to be installed. You can install these with `pip`:  
`pip install pandas numpy scanpy matplotlib`  

Once these libraries are installed, `di-SNE` can be installed with the following command:  
`python setup.py install`  

If the install was successful, the command `diSNE --help` should display a help message.  

# Usage  
`di-SNE` 

To run `diSNE` on a small test example:  

## Options  
The only required input to `diSNE` is a _ file. Any of the additional options below can be specified if desired:  
- `perplexity`: 
- `learning-rate`: 

# Contributors
This repository was generated by Nimrit Kaur, Tanvi Jain, and Kathryn Chen as a project for CSE 185, spring quarter 2024. We referenced 
